### ISSUE TITLE
Causes robolectric unable to proceed

### ISSUE THREAD

**Issue by ChrisZou** [rahatarmanahmed/CircularProgressView#21]
Hi, I was using this library and robolectric (3.0-rc3) for unit test. I found that `circularProgressView.startAnimation()` will prevent Robolectric from proceeding.

```Java
public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        CircularProgressView progressView = (CircularProgressView) findViewById(R.id.progress_view);
        progressView.startAnimation();
    }
}

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class MainActivityTest {
    @Test
    public void testMainActivity() {
        MainActivity mainActivity = Robolectric.setupActivity(MainActivity.class);
        Assert.assertNotNull(mainActivity);
    }
}
```

The program will stuck in `setupActivity()` and the assertion will never be reached. And if I comment out `progressView.startAnimation();`, then everything works fine. I don't know if its a problem of this library or of Robolectric, but would you take a look at this?

**Comment by kaoree** [comment]
please try write while(1){
} code in oncreate function, and see whether the  assertion will be reached or not;




------------------ 原始邮件 ------------------
发件人: "Chris";<notifications@github.com>;
发送时间: 2015年6月12日(星期五) 上午9:28
收件人: "rahatarmanahmed/CircularProgressView"<CircularProgressView@noreply.github.com>; 

主题: [CircularProgressView] Causes robolectric unable to proceed (#21)




Hi, I was using this library and robolectric (3.0-rc3) for unit test. I found that circularProgressView.startAnimation() will prevent Robolectric from proceeding.
 public class MainActivity extends AppCompatActivity {     @Override     protected void onCreate(Bundle savedInstanceState) {         super.onCreate(savedInstanceState);         setContentView(R.layout.activity_main);         CircularProgressView progressView = (CircularProgressView) findViewById(R.id.progress_view);         progressView.startAnimation();     } } @RunWith(RobolectricGradleTestRunner.class) @Config(constants = BuildConfig.class, sdk = 21) public class MainActivityTest {     @Test     public void testMainActivity() {         MainActivity mainActivity = Robolectric.setupActivity(MainActivity.class);         Assert.assertNotNull(mainActivity);     } }
 
The program will stuck in setupActivity() and the assertion will never be reached. And if I comment out progressView.startAnimation();, then everything works fine.

**Comment by ChrisZou** [comment]
@kaoree I don't understand. What is while(1){} code? I don't think java support using integer in while statement.

**Comment by kaoree** [comment]
sorry , true,not 1




------------------ 原始邮件 ------------------
发件人: "Chris";<notifications@github.com>;
发送时间: 2015年6月12日(星期五) 上午9:42
收件人: "rahatarmanahmed/CircularProgressView"<CircularProgressView@noreply.github.com>; 
抄送: "枫叶"<1052942629@qq.com>; 
主题: Re: [CircularProgressView] Causes robolectric unable to proceed (#21)




@kaoree I don't understand. What is while(1){} code? I don't think java support using integer in while statement.

**Comment by ChrisZou** [comment]
@kaoree Do you mean that CircularProgressView uses a while(true) loop and that's why the onCreate will never exit?

**Comment by rahatarmanahmed** [comment]
What happens if you use the `cpv_animAutostart` option in XML and remove `progressViewstartAnimation()`?

**Comment by ChrisZou** [comment]
@rahatarmanahmed It does no difference, the test will still stuck.

**Comment by rahatarmanahmed** [comment]
Can you try logging before and after `progressView.startAnimation()` to see if it actually completes? Also maybe a try-catch, in case Robolectric is swallowing the exception.

**Comment by ChrisZou** [comment]
@rahatarmanahmed I added some logs before and after `startAnimation`.

```Java
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        CircularProgressView progressView = (CircularProgressView) findViewById(R.id.progress_view);
        System.out.println("oncreate");
        try {
            progressView.startAnimation();
        } catch (Exception e) {
            System.out.println("exception happened when start animation");
            e.printStackTrace();
        }
        System.out.println("oncreate end");
    }
```

The "oncreate" and "oncreate end" messages were successfully logged and no exception was thrown.

**Comment by rahatarmanahmed** [comment]
How about after `startActivity()`? How do you know the assert statement isn't reached? `mainActivity` shouldn't be null so it won't say anything.

**Comment by ChrisZou** [comment]
I know because the test never stops, it just keeps running and running and nothing comes out. If it reaches then there would be a test result output and the running would end. But in this case, it doesn't.

**Comment by rahatarmanahmed** [comment]
Can you try debugging it and seeing what line it's hanging on? I think you should be able to pause it after it starts hanging.

**Comment by tjohn** [comment]
Hey so I think I know the issue here. With Robolectrics latest release the Animation classes are not faked or shadowed. This means that the actual animations are running. So you must call ShadowLooper.pauseMainLooper() before any animation.

**Comment by rahatarmanahmed** [comment]
I'm not at all familiar with Robolectric so what is the issue with this loop?

**Comment by tjohn** [comment]
I was seeing an infinite loop, I will do some more investigation and post
my findings.

Trevor

**Comment by rahatarmanahmed** [comment]
@alexandrudumbravan a pull request would be very welcome! I am no longer actively maintaining this library.

**Comment by lukaszkalnik** [comment]
@alexandrudumbravan thank you for this suggestion. I had similar problems with Espresso tests and your fix seems to have improved the situation. Do you want to submit a pull request or should I?

**Comment by alexandrudumbravan** [comment]
@lukaszkalnik go ahead. I don't think I would have the time for this

**Comment by rahatarmanahmed** [comment]
Just released v2.5.0, give it a few hours to show up. This includes @lukaszkalnik's PR, which should hopefully fix this issue.
